### PR TITLE
Generate manifest for examples page in viewer

### PR DESCRIPTION
**Pull request by jprochazk**
### What

Part of https://github.com/rerun-io/rerun/issues/3212

- Check that examples with a `thumbnail` have `thumbnail_dimensions` too
- Add `scripts/ci/thumbnails.py` script for checking/updating `thumbnail_dimensions`
- Add `build.rs` to `re_viewer`
  - Keeps `re_viewer/data/examples_manifest.json` up to date
  - The file is committed, and is only updated in rerun workspace/CI
  - URLs point to:
    - `version/nightly` on `main`/local builds
    - `version/x.y.z` on release branches
    - `commit/{short_hash}` on PRs

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3332) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/3332)
- [Docs preview](https://rerun.io/preview/1ba564f5afd6709dbe50ff411ba563b1d668caea/docs) 
- [Examples preview](https://rerun.io/preview/1ba564f5afd6709dbe50ff411ba563b1d668caea/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: .github/workflows/reusable_upload_web.yml [hunk 1]
```diff
@@ -100,21 +96,37 @@ jobs:
           parent: false
 
       - name: "Upload web-viewer (prerelease)"
-        if: inputs.MARK_PRERELEASE_FOR_MAINLINE
+        if: github.ref == 'refs/heads/master'
```
Comment (emilk) [inline .github/workflows/reusable_upload_web.yml:99]: `master`? Did you mean `main`?

File: .github/workflows/reusable_upload_web.yml [hunk 2]
```diff
@@ -100,21 +96,37 @@ jobs:
...
+          destination: "rerun-web-viewer/version/nightly"
+          parent: false
+
+      - name: "Upload RRD (nightly)"
+        if: github.ref == 'refs/heads/master'
+        uses: google-github-actions/upload-cloud-storage@v1
+        with:
+          path: "rrd"
+          destination: "rerun-example-rrd/version/nightly"
+          parent: false
```
Comment (emilk) [inline .github/workflows/reusable_upload_web.yml:128]: It is not run every night though, but on every merge into `main`. Maybe this should be called `trunk` instead?

Or do you envision this will be chnaged to nightly at some point, and so `nightly` is more of a promise of "at least once per day"?

File: Cargo.lock [hunk 3]
```diff
@@ -1765,7 +1765,7 @@ checksum = "79386fdcec5e0fde91b1a6a5bcd89677d1f9304f7f986b154a1b9109038854d9"
 dependencies = [
  "az",
  "bytemuck",
- "half 1.8.2",
+ "half 2.3.1",
```
Comment (emilk) [inline Cargo.lock:1768]: unrelated change?

File: crates/re_viewer/build.rs [hunk 4]
```diff
@@ -1,4 +1,212 @@
+use std::path::Path;
```
Comment (emilk) [inline crates/re_viewer/build.rs:1]: it would be helpful with a `//!` comment here explaining in short the purpose of this build-script

File: crates/re_viewer/build.rs [hunk 5]
```diff
@@ -1,4 +1,212 @@
...
+    }
+    re_build_tools::rerun_if_changed_or_doesnt_exist(MANIFEST_PATH);
+
+    if let Err(err) = write_examples_manifest() {
+        panic!("{err}");
+    }
+}
+
 fn main() {
     re_build_tools::rebuild_if_crate_changed("re_viewer");
```
Comment (emilk) [inline crates/re_viewer/build.rs:208]: This will re-run this build.rs script if anything in `re_viewer` or any of its dependancies changes. Isn't that a bit too often? I thought only the CI needed to run this at all? And it need to run it _always_,

File: scripts/ci/thumbnails.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
+#!/usr/bin/env python3
+
+"""Check or update example thumbnails."""
```
Comment (emilk) [inline scripts/ci/thumbnails.py:3]: it doesn't really update the thumbnails though, i.e. generate new images?

File: .github/workflows/reusable_upload_web.yml (hunk 2 above)
Comment (jprochazk) [reply .github/workflows/reusable_upload_web.yml:128]: Yes, we want to build all the examples, ideally once per day, but we're not there yet. `version/nightly` will eventually become that, and I figured there's no point in temporarily calling it something else.

File: Cargo.lock (hunk 3 above)
Comment (jprochazk) [reply Cargo.lock:1768]: I'm not sure why it decided to do this, I only added `serde_yaml` as a dev dependency. We're already on Rust 1.72, and `fixed` has its `half` version set as `>=1.8,<3`, so it seems harmless.

File: crates/re_viewer/build.rs (hunk 5 above)
Comment (jprochazk) [reply crates/re_viewer/build.rs:208]: The build script has to run locally so that we can commit the `examples_manifest.json` file when it changes. That way users can build the `re_viewer` crate without being in the workspace. We need to trigger a run of the build script on any change to the `examples` directory, but I don't know what the best way to detect if examples were added/changed/removed is. Always running it solves that, I think?

I tried profiling the build script, and maybe it won't turn into an issue. Before this PR, the script takes ~350ms to run. Calling `write_examples_manifest` added ~500us for me. On CI it will add a bit more, because it also has to run `git` a few times.

File: .github/workflows/reusable_checks.yml [hunk 7]
```diff
@@ -266,6 +266,7 @@ jobs:
       - name: Install dependencies
         run: |
           pip install gitignore_parser python-frontmatter
+          pip install -r ./scripts/ci/requirements.txt
```
Comment (emilk) [inline .github/workflows/reusable_checks.yml:269]: Neither of these should be needed if we would run the job on our `ci_docker` container

```
    container:
      image: rerunio/ci_docker:0.9
```

File: .github/workflows/reusable_upload_web.yml (hunk 2 above)
Comment (emilk) [reply .github/workflows/reusable_upload_web.yml]: Sounds good, but reading this code doesn't convey that - please add a comment!

File: crates/re_viewer/Cargo.toml [hunk 8]
```diff
@@ -113,3 +113,9 @@ web-sys = { version = "0.3.52", features = ["Window"] }
 
 [build-dependencies]
 re_build_tools.workspace = true
+
+# External
+serde = { version = "1", features = ["derive"] }
+serde_json = "1"
+serde_yaml = "=0.9.21"
```
Comment (emilk) [inline crates/re_viewer/Cargo.toml:120]: So, afaict this is to parse the markdown frontmatter, correct? Could we be using `toml` in there instead?

File: crates/re_viewer/build.rs (hunk 5 above)
Comment (emilk) [reply crates/re_viewer/build.rs:227]: Oh, I missed where I was.

> We need to trigger a run of the build script on any change to the examples directory, but I don't know what the best way to detect if examples were added/changed/removed is

For changed/removed, use `re_build_tools::rerun_if_changed` for each file you read. It does `println!("cargo:rerun-if-changed=…` internally.

Detecting added files… I don't know how to do.

File: crates/re_viewer/build.rs [hunk 9]
```diff
@@ -1,4 +1,231 @@
...
+            }
+            examples.push(Example { name, readme });
+        }
+    }
+    Ok(examples)
+}
+
+fn parse_frontmatter<P: AsRef<Path>>(path: P) -> Result<Option<Frontmatter>> {
+    let path = path.as_ref();
+    let content = std::fs::read_to_string(path)?;
```
Comment (emilk) [inline crates/re_viewer/build.rs:156]:
```suggestion
    re_build_tools::rerun_if_changed(path);
    let content = std::fs::read_to_string(path)?;
```

File: crates/re_viewer/build.rs (hunk 5 above)
Comment (emilk) [reply crates/re_viewer/build.rs:227]: > Always running it solves that, I think?

If you want to force the `build.rs` file to _always_ run (which you may want to on CI), there is an ugly trick: `println!("cargo:rerun-if-changed=path/to/file/that/doesnt/exist");`

File: examples/python/plots/README.md [hunk 10]
```diff
@@ -1,9 +1,11 @@
 ---
 title: Plots
 python: https://github.com/rerun-io/rerun/tree/latest/examples/python/plots/main.py
-thumbnail: https://static.rerun.io/ca0c72df93d70c79b0e640fb4b7c33cdc0bfe5f4_plots_480w.png
-description: "Demonstration of various plots and charts supported by Rerun."
 tags: [2d, plots, api-example]
+description: "Demonstration of various plots and charts supported by Rerun."
+thumbnail: https://static.rerun.io/ca0c72df93d70c79b0e640fb4b7c33cdc0bfe5f4_plots_480w.png
+thumbnail_dimensions: [480, 271]
+build_args: []
```
Comment (emilk) [inline examples/python/plots/README.md:8]: This line looks like it is redundant - surely the default is an empty list, so I can just remove it from the file, right?

Well, that is very reasonable, unless you very recently read this comment at the top of `crates/re_viewer/build.rs`:

> …only includes those with a `build_args` set in their `README.md` frontmatter.

Oh shit, `build_args` is load bearing!?

Much safer to make this explcit:

```suggestion
include_on_welcome_page: true
```

**Review by emilk — CHANGES_REQUESTED** [review]

File: crates/re_viewer/Cargo.toml (hunk 8 above)
Comment (jprochazk) [reply crates/re_viewer/Cargo.toml:120]: Yes, we'd also have to update everything else that reads the frontmatter. Currently that's at least landing and `build_demo_app.py`

File: .github/workflows/reusable_checks.yml (hunk 7 above)
Comment (jprochazk) [reply .github/workflows/reusable_checks.yml:269]: I think the docker container is a hack we should use as little as possible, because someone always has to remember to update it any time we add new dependencies. The job here takes 1 minute to run, and it runs in parallel to all the other checks.

File: crates/re_viewer/Cargo.toml (hunk 8 above)
Comment (emilk) [reply crates/re_viewer/Cargo.toml:120]: Please do 🙏 , but can be saved for another PR

File: crates/re_viewer/build.rs (hunk 5 above)
Comment (jprochazk) [reply crates/re_viewer/build.rs:229]: It should already always run on CI, because all the crates are built from scratch

File: crates/re_viewer/build.rs [hunk 11]
```diff
@@ -1,4 +1,233 @@
...
+                    Some(value) => value,
+                    None => bail!("{:?} is missing field {:?}", $e.name, stringify!($f)),
+                }
+            };
+        }
+
+        let thumbnail_dimensions = get!(example, thumbnail_dimensions);
+
+        Ok(Self {
+            title: get!(example, title),
```
Comment (emilk) [inline crates/re_viewer/build.rs:109]: Why use a macro to unwrap the `title: Option<String>` instead of using `title: String` directly?

**Review by emilk — APPROVED** [review]

File: crates/re_viewer/build.rs (hunk 11 above)
Comment (jprochazk) [reply crates/re_viewer/build.rs]: I think I just wanted the parsing from YAML to not fail early, so that I could check for the presence of `build_args`. I just defaulted all the fields instead.

File: .github/workflows/reusable_checks.yml (hunk 7 above)
Comment (jprochazk) [reply .github/workflows/reusable_checks.yml:271]: I wanted to try it to see how much of a speedup it would be, because I suspect it might actually be a slowdown, but:

> python3: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by python3)

Now I don't want to touch it in fear of what other horrors lie beyond. 🙂
